rust/worker/src/execution/operators/metadata_filtering.rs

```diff
@@ -35,75 +55,353 @@ impl MetadataFilteringOperator {
+                        metadata_segment_reader.u32_metadata_index_reader.as_ref(),
+                        &(*i as u32).into(),
+                    ),
+                    Float(f) => (
+                        metadata_segment_reader.f32_metadata_index_reader.as_ref(),
+                        &(*f as f32).into(),
+                    ),
+                    Str(s) => (
+                        metadata_segment_reader
+                            .string_metadata_index_reader
+                            .as_ref(),
+                        &s.as_str().into(),
+                    ),
+                };
+                if let Some(reader) = metadata_index_reader {
+                    match op {
+                        Equal => Ok(reader.get(key, kw).await?),
+                        GreaterThan => Ok(reader.gt(key, kw).await?),
+                        GreaterThanOrEqual => Ok(reader.gte(key, kw).await?),
+                        LessThan => Ok(reader.lt(key, kw).await?),
+                        LessThanOrEqual => Ok(reader.lte(key, kw).await?),
+                        _ => Err(MetadataFilteringError::InvalidInput),
+                    }
+                } else {
+                    Ok(RoaringBitmap::new())
+                }
+            }
+            Log(metadata_log_reader) => metadata_log_reader.get(key, val, op),
+        }
+    }
+}
+
+pub(crate) trait RoaringMetadataFilter<'me> {
+    async fn eval(
+        &'me self,
+        meta_provider: &MetadataProvider<'me>,
+    ) -> Result<SignedRoaringBitmap, MetadataFilteringError>;
+}
+
+impl<'me> RoaringMetadataFilter<'me> for Where {
+    async fn eval(
+        &'me self,
+        meta_provider: &MetadataProvider<'me>,
+    ) -> Result<SignedRoaringBitmap, MetadataFilteringError> {
+        use Where::*;
+        match self {
+            DirectWhereComparison(direct_comparison) => direct_comparison.eval(meta_provider).await,
+            DirectWhereDocumentComparison(direct_document_comparison) => {
+                direct_document_comparison.eval(meta_provider).await
+            }
+            WhereChildren(where_children) => Box::pin(where_children.eval(meta_provider)).await,
+        }
+    }
+}
+
+impl<'me> RoaringMetadataFilter<'me> for DirectWhereComparison {
+    async fn eval(
+        &'me self,
+        meta_provider: &MetadataProvider<'me>,
+    ) -> Result<SignedRoaringBitmap, MetadataFilteringError> {
+        use MetadataSetValue::*;
+        use PrimitiveOperator::*;
+        use SetOperator::*;
+        use SignedRoaringBitmap::*;
+        let result = match &self.comp {
+            WhereComparison::Primitive(primitive_operator, metadata_value) => {
+                match primitive_operator {
+                    NotEqual => Exclude(
+                        meta_provider
+                            .filter_by_metadata(&self.key, metadata_value, &Equal)
+                            .await?,
+                    ),
+                    _ => Include(
```

Comment:
nit: prefer enumerating all the cases here (even if code is duplicated). This ensure that in future when someone adds a new type of primitive_operator, the compiler forces them to add it here also otherwise it could be missed

Reply:
I expanded the case here